design/anoncreds.md

```diff
@@ -0,0 +1,507 @@
+    1. One needs to be able to get all REVOC_REG_DEFs created by the given Issuer DID for the given CRED_DEF.    
+
+### Technical goals
+* Define how Schemas, CreDefs and RevocRegDefs are identified (seqNo, UUID, primary key tuples?)
+* Define how to deal with Requirements 6.
+* Define Revocation-related transactions and necessary changes in indy-node.
+
+
+### Referencing Schema and CredDef in Credentials and Proofs
+* Schema is referenced by unique `SchemaUUID`.
+    * Created for each new Schema.
+    * This is different from Schema Issuer DID (DID used to send `SCHEMA` txn) which can be the same for 
+    any number of Schemas.
+* CredDef is referenced by unique `CredDefUUID`.
+    * Created for each new CredDef.
+    * This is different from CredDef Issuer DID (DID used to send `CRED_DEF` txn) which can be the same for 
+    any number of CredDef.
+* RevocRegDef and RevocReg are referenced by unique `RevocRegDefUUID`.
+    * Created for each new RevocRegDef.
+    * Used to update RevocReg.
+    * This is different from RevocRegDef Issuer DID (DID used to send `REVOC_REG_DEF` txn) which can be the same for 
+    any number of RevocRegDef.
+* Each `UUID` can be considered as a global UUID within the Ledger.
+The Ledger can guarantee that malicious party can not change/break existing entity 
+defined by a UUID by 
+and checking if there is an entity with the given UUID already existent, plus
+checking the ownership when modifying existing entities (only the issuer who created en entity with the given UUID can modify it). 
+
+      
+### How Prover and Verifier get keys for Credentials and Proofs
+* Proofs and credentials come with `schemaUUID`, `credDefUUID`, `revocDefUUID`.
+* Also there can be `issuanceTime` attribute in each credential (which can be disclosed in the proof).
+* Prover/Verifier looks up SCHEMA using `GET_SCHEMA(schemaUUID)` request to the ledger
+* Prover/Verifier looks up CRED_DEF using `GET_CRED_DEF(credDefUUID, issuanceTime)` request to the ledger
+* Prover/Verifier looks up REVOC_REG_DEF using `GET_REVOC_REG_DEF(revocDefUUID, issuanceTime)` request to the ledger
+* Prover looks up REVOC_REG to update the witness for the given `timestamp` 
+ using `GET_REVOC_REG(revocDefUUID, timestamp)` request to the ledger
+* Verifies looks up REVOC_REG to get accumulator value for the given `timestamp`  
+ using `GET_REVOC_REG(revocDefUUID, timestamp)` request to the ledger
+* Prover and Verifier should look up REVOC_REG by the same `timestamp` when generating and verifying the proof.
+
+### Timestamp Support in State
+
+We need to have a generic way to get the State at the given time.
+- State Trie allows to go to the past, that is given the root hash, get the state with this root.
+- We may have a mapping of each State update (timestamp) to the corresponding root.
+- We need to find a data structure that can help us to find the nearest state timestamp (to get the root) for the given
```

Comment:
Leveldb and Rocksdb keep the keys sorted so you can keep the timestamp as a key (part of key, use a custom comparator)  and the value for that key is the key in trie

Reply:
Yes. But we also need to be able to find the value for the timestamps which are not exact keys in the Leveldb.
So, let's assume that it contains keys `timeA` and  `timeB`, and we want to the get the value for `timeC`, so that `timeA < timeC < timeB`.